Db model for game workflow integration

Iris-Chensy opened the pull request:
This branch implements a complete database redesign to support the MemorySim narrative game workflow, integrating LLM generation, RAG retrieval. The design builds upon teammate @brynn-du's foundational work while adding comprehensive workflow support.

### Core Model Architecture

**Turn Model - Game State Management**
- Year as primary key for shared session gameplay where all users see identical content
- Added Question phase fields: question, question_generated_at for LLM-generated questions
- Added Description phase fields: description, description_generated_at for post-choice narratives
- Integrated RAG/LLM storage: query_text for search queries, context_block for retrieved knowledge
- Preserved teammate's image management: displayed_image_rel for turn-specific images
- Added user_choice field to track selected options (A/B/C/D)

**Option Model - Choice Management**
- Restructured to support exactly 4 options per turn (A, B, C, D) with uniqueness constraints
- Separated option_text for display and image_text for AI image generation
- Supports phased data population as LLM generates content

**ImageRender Model - Enhanced Pipeline**
- Built on teammate's original design with status tracking (pending/ready/failed)
- Use last_turn_image_rel fallback mechanism for failed generations
- Connected to Option model for per-choice image generation

**New Models**
- WorldBackground: Reusable story content across sessions
- DefaultQueryList: Initial keywords for year 2035 (simplified random selection)
- Session: Enhanced for multi-user shared gameplay

**Technical Improvements**
Database: Year-based primary key for fast lookup, unique constraints, proper foreign keys, optimized indexing

# Type of change
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [ ] Admin Interface: http://127.0.0.1:8000/admin/ (username: admin, password: admin123)

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

brynn-du-uq committed:
initial models

brynn-du-uq committed:
added a new field in Turn 'revealed_image_filename' and modified fields name in ImageRender to avoid confusion

brynn-du-uq committed:
name changes

Iris-Chensy committed:
Redesign database models for game workflow integration

- Redesigned Turn model with year as primary key for shared session gameplay
- Added comprehensive fields for Question and Description phases
- Integrated RAG and LLM interaction data storage (query_text, context_block)
- Restructured Option model to support 4 options per turn (A, B, C, D)
- Simplified DefaultQueryList for random keyword selection
- Added WorldBackground model for reusable story content
- Updated admin interface for better data visualization

brynn-du-uq commented on backend/shared/models.py:
```diff
@@ -0,0 +1,95 @@
+    Core model: stores all data for a specific turn/year in the game.
+    Uses year as primary key for easy lookup.
+    """
+    year = models.IntegerField(primary_key=True) # Year as primary key (e.g., 2035, 2036...)
```
.

Iris-Chensy replied:
For the primary key, I choose session+year instead of turn_id:
1. Year is a natural business primary key - Each year within a session is inherently unique
2. Higher code readability - Year has clear business meaning and context and easy to find last_year data
3. More intuitive frontend interaction - URLs and state management are simpler and more semantic
4. More efficient data queries - Direct year-based queries without additional mapping layers
5. Aligns with game logic - Players understand "2035's decision" better than "Turn 3's decision"

brynn-du-uq commented on backend/shared/models.py:
```diff
@@ -0,0 +1,95 @@
+    context_block = models.TextField(blank=True) # RAG retrieved context
+    
+    # User interaction (for shared session voting/consensus)
+    user_choice = models.CharField(max_length=1, blank=True) # A, B, C, or D - final chosen option
```
fk

brynn-du-uq commented on backend/shared/models.py:
```diff
@@ -0,0 +1,95 @@
+    label = models.CharField(max_length=1) # A, B, C, D
+    option_text = models.TextField(blank=True) # The actual option text displayed to user (may be empty initially)
+    image_text = models.TextField(blank=True) # LLM generated prompt for image generation (added later)
+
```
created_at = models.DateTimeField(auto_now_add=True)

brynn-du-uq commented on backend/shared/models.py:
```diff
@@ -0,0 +1,95 @@
+    option = models.ForeignKey(Option, on_delete=models.CASCADE, related_name='renders')
+    status = models.CharField(max_length=16, choices=STATUS_CHOICES, default='pending')
+    image_rel = models.CharField(max_length=512, blank=True)
+    last_turn_image_rel = models.CharField(max_length=512, blank=True) # image of the last turn, as a fallback
```
created_at = models.DateTimeField(auto_now_add=True)

brynn-du-uq replied:
remove last_turn_image_rel

Iris-Chensy committed:
update Turn model for multi-session support
- Use composite primary key (session, year) instead of single year
- Change user_choice from CharField to ForeignKey(Option)
- Support parallel game sessions with independent timelines

brynn-du-uq commented:
[Composite keys does not natively work in Django and it is likely to introduce errors and bugs](https://docs.djangoproject.com/en/5.2/topics/composite-primary-key/). If we want to track year in Turn maybe make that into a field, like turn_index, every time a question is generated create a Turn with the turn_index being the largest turn_index in the current session + 1, then we will be able to compute the current year based on this index and maybe make that into a field as well (e.g. current_year), instead of storing the raw year int like 2035.

brynn-du-uq commented on backend/shared/models.py:
```diff
@@ -0,0 +1,97 @@
+    Uses (session, year) as composite primary key to support multiple parallel sessions.
+    """
+    session = models.ForeignKey(Session, on_delete=models.CASCADE, related_name='turns')
+    year = models.IntegerField() # Year within this session (e.g., 2035, 2036...)
```
turn_index = models.PositiveSmallIntegerField() # 0,1,2...
when a new Turn is created, increment the index, then use the index to calculate the current year = base_year + turn_index, depending on how you will make use of the year. I suggest that we avoid using composite keys since it's not natively supported and we can just use turn_id to directly find a Turn, using a composite key just introduces extra complexity. For year track, it can just be a field.

Iris-Chensy replied:
Quick clarification on Turn model: My previous comment was misleading - ['session', 'year'] is actually a unique constraint, not a composite primary key. The Turn model still uses Django's standard [id] as the primary key. This design avoids Django's composite key complexities while ensuring each session has unique years, and allows flexible querying by either ID or session+year combination. Database migration completed successfully.

Iris-Chensy committed:
Fix Turn model: Use unique constraint instead of composite primary key

IanP96 commented on backend/shared/models.py:
```diff
@@ -0,0 +1,99 @@
+    Each turn has exactly 4 options (A, B, C, D) with corresponding image prompts.
+    """
+    turn = models.ForeignKey(Turn, on_delete=models.CASCADE, related_name='options')
+    label = models.CharField(max_length=1) # A, B, C, D
```
this field (label) could be an enum instead? since it is limited to A, B, C or D. WOuld only be a minor change but would help maintain data integrity

IanP96 reviewed (approved):
I can access the Django admin page and see the created models as required.
Could you please add instructions for how to add a user with the required username and password (or any username/password), probably in the README

brynn-du-uq committed:
fixed documentation
